src/controller.cpp

```diff
@@ -196,6 +193,130 @@ void controller::setup_dirs(bool silent) {
 
 	searchfile = strprintf::fmt("%s%shistory.search", config_dir, NEWSBEUTER_PATH_SEP);
 	cmdlinefile = strprintf::fmt("%s%shistory.cmdline", config_dir, NEWSBEUTER_PATH_SEP);
+
+	return first_run;
+}
+
+void copy_file(
+		const std::string& input_filepath,
+		const std::string& output_filepath)
+{
+	std::ifstream src(input_filepath, std::ios_base::binary);
+	std::ofstream dst(output_filepath, std::ios_base::binary);
+	dst << src.rdbuf();
+}
+
+bool controller::migrate_data_from_newsbeuter_xdg(const char* env_home, bool silent) {
+	const char* env_xdg_config = ::getenv("XDG_CONFIG_HOME");
+	std::string xdg_config_dir;
+	if (env_xdg_config) {
+		xdg_config_dir = env_xdg_config;
+	} else {
+		xdg_config_dir = env_home;
+		xdg_config_dir.append(NEWSBEUTER_PATH_SEP);
+		xdg_config_dir.append(".config");
+	}
+
+	const char* env_xdg_data = ::getenv("XDG_DATA_HOME");
+	std::string xdg_data_dir;
+	if (env_xdg_data) {
+		xdg_data_dir = env_xdg_data;
+	} else {
+		xdg_data_dir = env_home;
+		xdg_data_dir.append(NEWSBEUTER_PATH_SEP);
+		xdg_data_dir.append(".local");
+		xdg_data_dir.append(NEWSBEUTER_PATH_SEP);
+		xdg_data_dir.append("share");
+	}
+
+	const auto newsbeuter_config_dir
+		= xdg_config_dir + NEWSBEUTER_PATH_SEP + NEWSBEUTER_SUBDIR_XDG + NEWSBEUTER_PATH_SEP;
+	const auto newsbeuter_data_dir
+		= xdg_data_dir + NEWSBEUTER_PATH_SEP + NEWSBEUTER_SUBDIR_XDG + NEWSBEUTER_PATH_SEP;
+
+	const auto newsboat_config_dir
+		= xdg_config_dir + NEWSBEUTER_PATH_SEP + NEWSBOAT_SUBDIR_XDG + NEWSBEUTER_PATH_SEP;
+	const auto newsboat_data_dir
+		= xdg_data_dir + NEWSBEUTER_PATH_SEP + NEWSBOAT_SUBDIR_XDG + NEWSBEUTER_PATH_SEP;
+
+	bool newsbeuter_config_dir_exists = 0 == access(newsbeuter_config_dir.c_str(), R_OK | X_OK);
+
+	if (! newsbeuter_config_dir_exists) {
+		return false;
+	}
+
+	utils::mkdir_parents(newsboat_config_dir, 0700);
+	utils::mkdir_parents(newsboat_data_dir, 0700);
+
+	if (! silent) {
+		std::cerr << "Migrating configs and data from Newsbeuter..." << std::endl;
+	}
+
+	/* in config */
+	copy_file(newsbeuter_config_dir + "urls", newsboat_config_dir + "urls");
+	copy_file(newsbeuter_config_dir + "config", newsboat_config_dir + "config");
```

Comment:
I'm not sure about blindily copying files, it could end up overwriting user data. Maybe just print a warning and abort copying that file if it already exists in the destination?

Additionally if there's a permission error it simply dumps an empty file without any warnings which is in my opinion a bad practice.

Reply:
This concern should be irrelevant now because now I'm always checking if the target dir exists before copying anything into it.

Reply:
> Additionally if there's a permission error it simply dumps an empty file without any warnings which is in my opinion a bad practice.

Hmm, maybe this one *isn't* irrelevant though. You're thinking of the case when, say, `~/.config/newsbeuter/urls` isn't readable, even though the directory itself is?

Reply:
One concern was about overwriting existing data but that was indeed resolved by checking whether the directories already exist or not.

The second is just a passing remark that if we don't have read permissions for a file it's just copied as empty instead of printing a warning/error. But I guess that's a rare enough occurrence to not bother with a check for that.